applications/CompressiblePotentialFlowApplication/custom_processes/define_3d_wake_process.cpp

```diff
@@ -0,0 +1,888 @@
+
+    return {p_node1->Id(), p_node2->Id(), p_node3->Id(), p_node4->Id()};
+}
+
+double Define3DWakeProcess::ComputeFaceNormalProjectionToWakeNormal(
+    const array_1d<double, 3>& rCoordinates1,
+    const array_1d<double, 3>& rCoordinates2,
+    const array_1d<double, 3>& rCoordinates3,
+    const array_1d<double, 3>& rCoordinates4) const
+{
+    const auto& side1 = rCoordinates2 - rCoordinates1;
+    const auto& side2 = rCoordinates3 - rCoordinates1;
+    array_1d<double, 3> face_normal = ZeroVector(3);
+    MathUtils<double>::CrossProduct(face_normal, side1, side2);
+    return inner_prod(face_normal, mWakeNormal);
+}
+
+void Define3DWakeProcess::CreateWakeSurfaceElements(const double normal_projection,
+                                                    IndexType& rElement_index,
+                                                    const std::vector<ModelPart::IndexType>& rNodes_ids,
+                                                    const Properties::Pointer pElemProp) const
+{
+    if(normal_projection > 0.0){
+        const std::vector<ModelPart::IndexType> elemNodes1{rNodes_ids[0], rNodes_ids[1], rNodes_ids[2]};
+        const std::vector<ModelPart::IndexType> elemNodes2{rNodes_ids[0], rNodes_ids[3], rNodes_ids[2]};
+        mrStlWakeModelPart.CreateNewElement("Element3D3N", ++rElement_index, elemNodes1, pElemProp);
+        mrStlWakeModelPart.CreateNewElement("Element3D3N", ++rElement_index, elemNodes2, pElemProp);
+    }
+    else{
+        const std::vector<ModelPart::IndexType> elemNodes1{rNodes_ids[0], rNodes_ids[2], rNodes_ids[1]};
+        const std::vector<ModelPart::IndexType> elemNodes2{rNodes_ids[0], rNodes_ids[2], rNodes_ids[3]};
+        mrStlWakeModelPart.CreateNewElement("Element3D3N", ++rElement_index, elemNodes1, pElemProp);
+        mrStlWakeModelPart.CreateNewElement("Element3D3N", ++rElement_index, elemNodes2, pElemProp);
+    }
+}
+
+// This function checks which elements are cut by the wake and marks them as
+// wake elements
+void Define3DWakeProcess::MarkWakeElements() const
+{
+    KRATOS_TRY;
+    KRATOS_INFO("MarkWakeElements") << "...Selecting wake elements..." << std::endl;
+    ModelPart& root_model_part = mrBodyModelPart.GetRootModelPart();
+
+    // Variable to store element ids
+    std::vector<std::size_t> wake_elements_ordered_ids;
+    std::vector<std::size_t> trailing_edge_elements_ordered_ids;
+
+    // TODO: substitute with CalculateDiscontinuousDistanceToSkinProcess
+    const auto start_time(std::chrono::steady_clock::now());
+    CalculateDistanceToSkinProcess<3> distance_calculator(root_model_part, mrStlWakeModelPart);
+    distance_calculator.Execute();
+    const auto end_time = std::chrono::duration_cast<std::chrono::duration<double>>(
+                              std::chrono::steady_clock::now() - start_time)
+                              .count();
+    KRATOS_INFO("MarkWakeElements") << " distance_calculator took " << end_time
```

Comment:
better use `utilities/builtin_timer`

also better protect this with an echo_level